src/System.Private.CoreLib/shared/System/Runtime/InteropServices/Marshal.cs

```diff
@@ -0,0 +1,718 @@
+            if (nb == 0)
+            {
+                return string.Empty;
+            }
+
+            return new string((sbyte*)ptr);
+        }
+
+        public static unsafe string PtrToStringAnsi(IntPtr ptr, int len)
+        {
+            if (ptr == IntPtr.Zero)
+            {
+                throw new ArgumentNullException(nameof(ptr));
+            }
+            if (len < 0)
+            {
+                throw new ArgumentException(null, nameof(len));
+            }
+
+            return new string((sbyte*)ptr, 0, len);
+        }
+
+        public static unsafe string PtrToStringUni(IntPtr ptr, int len)
+        {
+            if (ptr == IntPtr.Zero)
+            {
+                throw new ArgumentNullException(nameof(ptr));
+            }
+            if (len < 0)
+            {
+                throw new ArgumentException(SR.ArgumentOutOfRange_NeedNonNegNum, nameof(len));
+            }
+
+            return new string((char*)ptr, 0, len);
+        }
+
+        public static string PtrToStringAuto(IntPtr ptr, int len)
+        {
+            // Ansi platforms are no longer supported
+            return PtrToStringUni(ptr, len);
+        }
+
+        public static unsafe string PtrToStringUni(IntPtr ptr)
+        {
+            if (IntPtr.Zero == ptr)
+            {
+                return null;
+            }
+            else if (IsWin32Atom(ptr))
+            {
+                return null;
+            }
+
+            return new string((char*)ptr);
+        }
+
+        public static string PtrToStringAuto(IntPtr ptr)
+        {
+            // Ansi platforms are no longer supported
+            return PtrToStringUni(ptr);
+        }
+
+        public static unsafe string PtrToStringUTF8(IntPtr ptr)
+        {
+            if (IntPtr.Zero == ptr)
+            {
+                return null;
+            }
+
+            int nbBytes = string.strlen((byte*)ptr);
+            return PtrToStringUTF8(ptr, nbBytes);
+        }
+
+        public static unsafe string PtrToStringUTF8(IntPtr ptr, int byteLen)
+        {
+            if (byteLen < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(byteLen), SR.ArgumentOutOfRange_NeedNonNegNum);
+            }
+            else if (IntPtr.Zero == ptr)
+            {
+                return null;
+            }
+            else if (IsWin32Atom(ptr))
+            {
+                return null;
+            }
+            else if (byteLen == 0)
+            {
+                return string.Empty;
+            }
+
+            return Encoding.UTF8.GetString((byte*)ptr, byteLen);
+        }
+
+        public static int SizeOf(object structure)
+        {
+            if (structure == null)
+            {
+                throw new ArgumentNullException(nameof(structure));
+            }
+
+            return SizeOfHelper(structure.GetType(), true);
+        }
+
+        public static int SizeOf<T>(T structure) => SizeOf((object)structure);
```

Comment:
This seems surprising and unnecessarily costly.  Seems like this could instead be:
```C#
public static int SizeOf<T>(T structure)
{
    if (structure == null)
    {
        throw new ArgumentNullException(nameof(structure));
    }

    return SizeOfHelper(structure.GetType(), throwIfNotMarshalable: true);
}
```
?